[[ Bug 21503 ]] Tree widget - prevent value removal on key addition

bwmilby opened the pull request:
Current operation of the tree widget will delete the value stored in a key
if the "+" is clicked to add a new sub-key and it is not already an array.

This change alters the behavior such that when the "+" is clicked, if the current key is not already an array, then a dialog will ask the user for how to proceed.  The options are to `move` the existing value to the new sub-key that is created,`replace` the value with an empty array, or `cancel` the operation.

bwmilby committed:
[[ Bug 21503 ]] Tree widget - prevent value removal on key addition

Current operation of the tree widget will delete the value stored in a key
if the "+" is clicked to add a new sub-key and it is not already an array.

This change alters the behavior such that when the "+" is clicked, if the
current key is not already an array, then the existing value is moved to
the new sub-key that is created.

montegoulding commented on extensions/widgets/treeview/treeview.lcb:
```diff
@@ -1511,7 +1511,7 @@ private handler addKey(in pListElt as Integer, in pPath as List, in pLevel as In
 		variable tArray as Array
 		put the empty array into tArray	
-		put "" into tArray[1 formatted as string]
+		put tElement into tArray[1 formatted as string]
```
@bwmilby I wonder if rather than move the element which might be hard to recover from it should check if `tElement` is empty or not and post an error message if it's not. That way script can handle the message and ask the user if they want to move the current value to an array element. It could be they just clicked `+` on the wrong line and would like to back out.

montegoulding replied:
We could I guess add the reverse of this behavior so it's easy to back out but it's not all that intuitive to click the bin on an array element in an array with one key and have the value clobber the array.

bwmilby replied:
I considered the error route (similar to the delete message) but thought this would be less intrusive for applications.  The problem about the reverse is you have a key and a value to deal with so you will lose one anyway.

montegoulding replied:
Hmm... yeah the way the delete message is done would work rather than a callback. I imagine the main choice people would want to make is between canceling and replacing the element with an array. Using the current element value as array key 1 is possibly more of an edge case... not sure. I guess it could be a third button on the dialog...

```
answer warning "The array element at path [" & <path> & "] is not empty. Would you like to replace it with an empty array or move it to the first element?" with "Cancel" or "Replace" or "Move"
```

montegoulding replied:
Great thanks @bwmilby

bwmilby committed:
[[ Bug 21503 ]] Tree widget - prevent value removal on key addition

Adjust code to display a dialog box to allow the user to choose how to handle
the situation where a new array could remove an existing value.

montegoulding commented:
@livecode-vulcan review ok b70a409

livecode-vulcan commented:
💙 review by @montegoulding ok b70a409911ffaeaa69c941f5da3e56c2fbde7cfc

livecode-vulcan commented:
😎 test success b70a409911ffaeaa69c941f5da3e56c2fbde7cfc

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/62)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/54)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/54)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/54)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/54)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1576)
* try-community-universal-ios-iphoneos11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.4/builds/103)
* try-community-universal-ios-iphonesimulator11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.4/builds/103)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1606)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1600)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1596)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1633)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1573)
